### PR TITLE
feat: Generate readmes

### PR DESCRIPTION
**Pull request by nalanj**
## Describe your changes

Adds a script `scripts/generate-script-readmes.ts` that generates and updates readmes dynamically to keep them in sync with nango.yaml for each script.

The way this works is that it does a very light-weight markdown parse based just on the header delimited sections and then transforms our known sections. The first handful of sections are required and kept up to date, and then anyone can add additional sections after those without the additional sections being altered.

This is a long PR because of all of the generated readmes. The critical file is `scripts/generate-script-readmes.ts`.

This runs in a pre-commit hook per @TBonnin's recommendation.

## Checklist before requesting a review (skip if just adding/editing APIs & templates)

-   [ ] I added tests, otherwise the reason is:
-   [ ] External API requests have `retries`
-   [ ] Pagination is used where appropriate
-   [ ] The built in `nango.paginate` call is used instead of a `while (true)` loop
-   [ ] Third party requests are NOT parallelized (this can cause issues with rate limits)
-   [ ] If a sync requires metadata the `nango.yaml` has `auto_start: false`
-   [ ] If the sync is a `full` sync then `track_deletes: true` is set
-   [ ] I followed the best practices and guidelines from the [Writing Integration Scripts](/NangoHQ/integration-templates/blob/main/WRITING_INTEGRATION_SCRIPTS.md) doc

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -11,3 +11,4 @@ globals.d.ts
 integrations/linkedin/proxy/video-upload.ts
 integrations/linkedin/nango-integrations/linkedin/proxy/video-upload.ts
 integrations/*-sandbox
+scripts/generate-script-readmes.ts
```
Comment (bodinsamuel) [inline .eslintignore:14]: Got too annoyed with eslint? It's not my repo but it's probably better to not flat out ignore everything :D

File: integrations/aircall-basic/actions/create-user.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  "email": "<string>",
+  "firstName": "<string>",
+  "lastName": "<string>"
+}
+```
+
+## Changelog
+
+- [Script History](https://github.com/NangoHQ/integration-templates/commits/main/integrations/aircall-basic/actions/create-user.ts)
+-- [ReadMe History](https://github.com/NangoHQ/integration-templates/commits/main/integrations/aircall-basic/actions/create-user.md)
```
Comment (bodinsamuel) [inline integrations/aircall-basic/actions/create-user.md:47]: Not sure it's common to say `ReadMe` plus it's very git. I guess the point is to share this to customers no?

**Comment by TBonnin** [comment]
> This isn't run in actions yet - I'll knock that out in an upcoming PR, but wanted to make sure we're generating good READMEs before we push forward.

I would do it as a pre-commit script so the output is actually reviewable

File: scripts/generate-script-readmes.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+// update sections to include latest content
+function updateReadme(
+    sections: MarkdownSections,
+    scriptName: string,
+    scriptPath: string,
+    endpointType: string,
+    scriptConfig: any,
+    models: any
+): MarkdownSections {
```
Comment (bodinsamuel) [inline scripts/generate-script-readmes.ts:74]: Why not splitting the file in two (with a well know pattern) and just always update everything. Seems a bit simpler, especially if you add/rename/remove section

Comment (nalanj) [reply scripts/generate-script-readmes.ts:83]: @bodinsamuel can you say more?

Comment (nalanj) [reply scripts/generate-script-readmes.ts]: See a381ed0. Is that more like what you're thinking?

Comment (bodinsamuel) [reply scripts/generate-script-readmes.ts]: seems much easier to maintain no? 👍🏻

File: .husky/pre-commit [hunk 4]
```diff
@@ -5,4 +5,8 @@ node ./scripts/flows.js
 npm run generate:tests --pre_commit=true
 git add ./integrations/*/tests/*.test.ts
 git add ./flows.yaml
+
+npm run generate:readmes
+git add ./integrations/**/*.md
+
```
Comment (bodinsamuel) [inline .husky/pre-commit:11]: If you want to commit in this repo you must be patient ahah

File: integrations/aircall-basic/actions/create-user.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
+# Create User
+
+## General Information
+
+- **Description:** Creates a user in Aircall.
+- **Version:** 0.0.1
+- **Group:** Others
+- **Scopes:**: _None_
```
Comment (bodinsamuel) [inline integrations/aircall-basic/actions/create-user.md:8]: Double `:`

File: integrations/aircall-basic/actions/create-user.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+- **Endpoint Type:** Action
```
Comment (bodinsamuel) [inline integrations/aircall-basic/actions/create-user.md:9]: Not sure we advertise this anymore

File: integrations/airtable/syncs/bases.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# Bases
+
+## General Information
+
+- **Description:** List all bases
+- **Version:** 0.0.1
+- **Group:** Others
+- **Scopes:**: schema.bases:read
```
Comment (bodinsamuel) [inline integrations/airtable/syncs/bases.md:8]: needs backtick 
```suggestion
- **Scopes:**: `schema.bases:read`
```

File: integrations/airtable/syncs/bases.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+- **Endpoint Type:** Sync
+- **Code:** [🔗](https://github.com/NangoHQ/integration-templates/tree/main/integrations/airtable/syncs/bases.ts)
```
Comment (bodinsamuel) [inline integrations/airtable/syncs/bases.md:10]: might be clearer to display `[github.com](...)`

File: integrations/intercom/syncs/conversations.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+- **Endpoint Type:** Sync
+- **Code:** [🔗](https://github.com/NangoHQ/integration-templates/tree/main/integrations/intercom/syncs/conversations.ts)
+
+
+## Endpoint Reference
+
+### Request Endpoint
+
+- **Path:** `undefined`
+- **Method:** `GET`
```
Comment (bodinsamuel) [inline integrations/intercom/syncs/conversations.md:19]: it's broken when scripts have multiple endpoints

File: .husky/pre-commit (hunk 4 above)
Comment (nalanj) [reply .husky/pre-commit:11]: Yeah, this was why I wanted it to be in an Action originally. In general I'm not sure adding pre-commit as a place for logic is a healthy practice.

File: scripts/generate-script-readmes.ts [hunk 10]
```diff
@@ -0,0 +1,182 @@
+/* eslint-disable @nangohq/custom-integrations-linting/no-try-catch-unless-explicitly-allowed */
```
[the remark this answers is not in the document]
Comment (nalanj) [reply scripts/generate-script-readmes.ts]: Changed in 03e188e. Originally there was discussion of us also generating readmes per integration, so I named it something mapping to scripts, but I'm pretty sure that discussion ended with us not doing so yet.

File: integrations/aircall-basic/actions/create-user.md (hunk 6 above)
Comment (nalanj) [reply integrations/aircall-basic/actions/create-user.md:9]: It was in the example readme I was given. I'll check in with @bastienbeurier about it.

**Comment by nalanj** [comment]
@bodinsamuel format came from here: https://linear.app/nango/issue/NAN-2321/integration-readme#comment-234a0e18

File: scripts/generate-script-readmes.ts (hunk 3 above)
Comment (nalanj) [reply scripts/generate-script-readmes.ts]: Yeah, previously the issue called for it to maintain these section titles while allowing human edits in between, but I think where this ended up we're set to just treat it as two big blocks.

File: integrations/intercom/syncs/conversations.md (hunk 9 above)
Comment (nalanj) [reply integrations/intercom/syncs/conversations.md:19]: Ooh, thanks. From spot checks I didn't realize that was possible.

File: integrations/aircall-basic/actions/create-user.md (hunk 5 above)
Comment (nalanj) [reply integrations/aircall-basic/actions/create-user.md]: Fixed in 3a57a79

File: integrations/airtable/syncs/bases.md (hunk 8 above)
Comment (nalanj) [reply integrations/airtable/syncs/bases.md]: Fixed in 3df2277a

File: integrations/aircall-basic/actions/create-user.md (hunk 6 above)
Comment (bastienbeurier) [reply integrations/aircall-basic/actions/create-user.md:9]: I think it's worth keeping. It's something we show in the dashboard as well.

**Review by bastienbeurier — APPROVED** [review]

File: integrations/intercom/syncs/conversations.md (hunk 9 above)
Comment (nalanj) [reply integrations/intercom/syncs/conversations.md:19]: See e0cf4dc9 for the fix based on our discussion. Also logged https://linear.app/nango/issue/NAN-2365/support-multiple-endpoints-for-readme-generation so we can come back and updated the format for multi-endpoint readmes.

File: scripts/generate-readmes.ts [hunk 11]
```diff
@@ -0,0 +1,184 @@
+/* eslint-disable @nangohq/custom-integrations-linting/no-try-catch-unless-explicitly-allowed */
+/* eslint-disable no-console */
+import { promises as fs } from 'fs';
+import { load as loadYaml } from 'js-yaml';
+
+const divider = '<!-- END  GENERATED CONTENT -->';
```
Comment (TBonnin) [inline scripts/generate-readmes.ts:6]: there is a `<!-- END  GENERATED CONTENT -->` tag but not a beginning one. Are we assuming it always starts at the top of the file?

Comment (nalanj) [reply scripts/generate-readmes.ts:6]: Yeah, I can toss one up top.

File: integrations/aircall-basic/actions/create-user.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
+# Create User
+
+## General Information
+
+- **Description:** Creates a user in Aircall.
+- **Version:** 0.0.1
+- **Group:** Others
+- **Scopes:** `undefined`
```
Comment (bodinsamuel) [inline integrations/aircall-basic/actions/create-user.md:8]: You still have undefined

Comment (nalanj) [reply integrations/aircall-basic/actions/create-user.md]: 🤬 OK, fixed in 7be48735

File: scripts/generate-readmes.ts (hunk 11 above)
Comment (nalanj) [reply scripts/generate-readmes.ts:6]: Fixed in 536a35c

**Review by bodinsamuel — APPROVED** [review]

**Review by khaliqgant — APPROVED** [review]

**Pull request merged**